pkg/querier/querier.go

```diff
@@ -69,19 +67,18 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.StoreGatewayClient.RegisterFlagsWithPrefix("querier.store-gateway-client", f)
 	f.BoolVar(&cfg.Iterators, "querier.iterators", false, "Use iterators to execute query, as opposed to fully materialising the series in memory.")
 	f.BoolVar(&cfg.BatchIterators, "querier.batch-iterators", true, "Use batch iterators to execute query, as opposed to fully materialising the series in memory.  Takes precedent over the -querier.iterators flag.")
-	f.DurationVar(&cfg.QueryIngestersWithin, queryIngestersWithinFlag, 13*time.Hour, "Maximum lookback beyond which queries are not sent to ingester. 0 means all queries are sent to ingester.")
 	f.DurationVar(&cfg.MaxQueryIntoFuture, "querier.max-query-into-future", 10*time.Minute, "Maximum duration into the future you can query. 0 to disable.")
 	f.DurationVar(&cfg.QueryStoreAfter, queryStoreAfterFlag, 12*time.Hour, "The time after which a metric should be queried from storage and not just ingesters. 0 means all queries are sent to store. If this option is enabled, the time range of the query sent to the store-gateway will be manipulated to ensure the query end is not more recent than 'now - query-store-after'.")
-	f.BoolVar(&cfg.ShuffleShardingIngestersEnabled, "querier.shuffle-sharding-ingesters-enabled", true, fmt.Sprintf("Fetch in-memory series from the minimum set of required ingesters, selecting only ingesters which may have received series since -%s. If this setting is false or -%s is '0', queriers always query all ingesters (ingesters shuffle sharding on read path is disabled).", queryIngestersWithinFlag, queryIngestersWithinFlag))
+	f.BoolVar(&cfg.ShuffleShardingIngestersEnabled, "querier.shuffle-sharding-ingesters-enabled", true, fmt.Sprintf("Fetch in-memory series from the minimum set of required ingesters, selecting only ingesters which may have received series since -%s. If this setting is false or -%s is '0', queriers always query all ingesters (ingesters shuffle sharding on read path is disabled).", validation.QueryIngestersWithinFlag, validation.QueryIngestersWithinFlag))
 
 	cfg.EngineConfig.RegisterFlags(f)
 }
 
 // Validate the config
-func (cfg *Config) Validate() error {
+func (cfg *Config) Validate(limits validation.Limits) error {
 	// Ensure the config wont create a situation where no queriers are returned.
-	if cfg.QueryIngestersWithin != 0 && cfg.QueryStoreAfter != 0 {
-		if cfg.QueryStoreAfter >= cfg.QueryIngestersWithin {
+	if limits.QueryIngestersWithin != 0 && cfg.QueryStoreAfter != 0 {
+		if cfg.QueryStoreAfter >= time.Duration(limits.QueryIngestersWithin) {
 			return errBadLookbackConfigs
 		}
 	}
```

Comment:
I'm unsure that this is how the `QueryIngestersWithin` limit should be validated. This will only validate the default limit once at startup. If there's a tenant override, it won't be validated and we might have a gap where we don't query from anywhere. 

I did it this way as it's consistent with how limits are validated for other components e.g. the ruler
https://github.com/grafana/mimir/blob/5ac2d2686b3a3e6b5c87c985cca5ace6970b2197/pkg/ruler/ruler.go#L120-L129

I could update this so tenant overrides are also validated. That would require us to pass in validation methods when we read updated tenant overrides.

Reply:
I believe doing the actual validation and erroring when reading the override as you say would be best, but here are some ideas/alternatives if that gets tricky:

1. Enforce and validate that `limits.QueryIngestersWithin` <= `limits[user].QueryIngestersWithin` (supposing that's easier to validate)
2. If it does happen that such gap exists and a query that falls there arrives, log an error and query all (or none) sources (don't really know the impact that querying all sources can have in performance)

Reply:
@pstibrany do you have any thoughts on this?

Reply:
Problem with checking `limits[user].QueryIngestersWithin` against `cfg.QueryStoreAfter` during query time is that user can't really do much about it... we need to let operators know.

Perhaps the best way to achieve that is to return 500, so that it counts towards error budget SLO and user doesn't accidentally get wrong results if `QueryStoreAfter >= QueryIngestersWithin` (because it means there is a gap that we don't query).

Reply:
An alternative may be moving this validation to the Limits and then not updating the limits if this validation fail. We have an alert firing if the runtime config is invalid (you may push an invalid YAML for example).

Reply:
I've gone for Marco's suggestion, adding a validation check to run every time runtime config is loaded: https://github.com/grafana/mimir/pull/4287/commits/9121057221159686af92eca7ef982713c43b496d. 
The load function will now call a `ValidateLimits()` function for the mimir `Config` struct, so each tenant's limits are validated. The approach I went for is a bit more generic than necessary for just `QueryIngestersWithin`. That's because there are other limits (e.g. `limits.RulerTenantShardSize`) that are currently only validated on startup but actually should be validated every time configs are loaded. I think we can update those in a future PR.